From 9590660dfe48e3d193ed7c9ef049ce6b768f64b8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 26 Sep 2023 16:34:16 +0200
Subject: [PATCH 01/17] makedep: Allow using an external wine object dir.

CW-Bug-Id: #22729
(cherry picked from commit ad57149a13295c31b2f01b423e9f3ad206ec39f4)
---
 tools/makedep.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/tools/makedep.c b/tools/makedep.c
index 9b3eb0623d1..a7b85faf5be 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -143,6 +143,7 @@ static struct strarray subdirs;
 static struct strarray delay_import_libs;
 static struct strarray top_install[NB_INSTALL_RULES];
 static const char *root_src_dir;
+static const char *root_obj_dir;
 static const char *tools_dir;
 static const char *tools_ext;
 static const char *wine64_dir;
@@ -737,6 +738,16 @@ static char *root_src_dir_path( const char *path )
 }
 
 
+/*******************************************************************
+ *         root_obj_dir_path
+ */
+static char *root_obj_dir_path( const char *path )
+{
+    if (!root_obj_dir) return (char *)path;
+    return concat_paths( root_obj_dir, path );
+}
+
+
 /*******************************************************************
  *         tools_dir_path
  */
@@ -2857,7 +2868,7 @@ static const char *cmd_prefix( const char *cmd )
 static void output_winegcc_command( struct makefile *make, unsigned int arch )
 {
     output( "\t%s%s -o $@", cmd_prefix( "CCLD" ), winegcc );
-    output_filename( "--wine-objdir ." );
+    output_filename( strmake( "--wine-objdir %s", root_obj_dir_path( "." ) ) );
     if (tools_dir)
     {
         output_filename( "--winebuild" );
@@ -4966,7 +4977,7 @@ static void load_sources( struct makefile *make )
         strarray_add( &make->include_args, strmake( "-I%s", make->src_dir ));
     if (make->parent_dir)
         strarray_add( &make->include_args, strmake( "-I%s", src_dir_path( make, make->parent_dir )));
-    strarray_add( &make->include_args, "-Iinclude" );
+    strarray_add( &make->include_args, strmake( "-I%s", root_obj_dir_path( "include" ) ) );
     if (root_src_dir) strarray_add( &make->include_args, strmake( "-I%s", root_src_dir_path( "include" )));
 
     list_init( &make->sources );
@@ -5101,6 +5112,7 @@ int main( int argc, char *argv[] )
         top_install[i] = get_expanded_make_var_array( top_makefile, strmake( "TOP_%s", install_variables[i] ));
 
     root_src_dir       = get_expanded_make_variable( top_makefile, "srcdir" );
+    root_obj_dir       = get_expanded_make_variable( top_makefile, "objdir" );
     tools_dir          = get_expanded_make_variable( top_makefile, "toolsdir" );
     tools_ext          = get_expanded_make_variable( top_makefile, "toolsext" );
     wine64_dir         = get_expanded_make_variable( top_makefile, "wine64dir" );
-- 
2.55.0

